drivers/qemu/driver.go

```diff
@@ -0,0 +1,647 @@
+	}
+
+	// Add pass through arguments to qemu executable. A user can specify
+	// these arguments in driver task configuration. These arguments are
+	// passed directly to the qemu driver as command line options.
+	// For example, args = [ "-nodefconfig", "-nodefaults" ]
+	// This will allow a VM with embedded configuration to boot successfully.
+	args = append(args, driverConfig.Args...)
+
+	// Check the Resources required Networks to add port mappings. If no resources
+	// are required, we assume the VM is a purely compute job and does not require
+	// the outside world to be able to reach it. VMs ran without port mappings can
+	// still reach out to the world, but without port mappings it is effectively
+	// firewalled
+	protocols := []string{"udp", "tcp"}
+	if len(cfg.Resources.NomadResources.Networks) > 0 {
+		// Loop through the port map and construct the hostfwd string, to map
+		// reserved ports to the ports listenting in the VM
+		// Ex: hostfwd=tcp::22000-:22,hostfwd=tcp::80-:8080
+		var forwarding []string
+		taskPorts := cfg.Resources.NomadResources.Networks[0].PortLabels()
+		for label, guest := range driverConfig.PortMap {
+			host, ok := taskPorts[label]
+			if !ok {
+				return nil, nil, fmt.Errorf("Unknown port label %q", label)
+			}
+
+			for _, p := range protocols {
+				forwarding = append(forwarding, fmt.Sprintf("hostfwd=%s::%d-:%d", p, host, guest))
+			}
+		}
+
+		if len(forwarding) != 0 {
+			args = append(args,
+				"-netdev",
+				fmt.Sprintf("user,id=user.0,%s", strings.Join(forwarding, ",")),
+				"-device", "virtio-net,netdev=user.0",
+			)
+		}
+	}
+
+	// If using KVM, add optimization args
+	if accelerator == "kvm" {
+		if runtime.GOOS == "windows" {
+			return nil, nil, errors.New("KVM accelerator is unsupported on the Windows platform")
+		}
+		args = append(args,
+			"-enable-kvm",
+			"-cpu", "host",
+			// Do we have cores information available to the Driver?
+			// "-smp", fmt.Sprintf("%d", cores),
+		)
+	}
+	d.logger.Debug("starting QemuVM command ", "args", strings.Join(args, " "))
+
+	pluginLogFile := filepath.Join(cfg.TaskDir().Dir, fmt.Sprintf("%s-executor.out", cfg.Name))
+	executorConfig := &dstructs.ExecutorConfig{
+		LogFile:  pluginLogFile,
+		LogLevel: "debug",
+	}
+
+	// TODO: best way to pass port ranges in from client config
+	execImpl, pluginClient, err := utils.CreateExecutor(os.Stderr, hclog.Debug, 14000, 14512, executorConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	execCmd := &executor.ExecCommand{
+		Cmd:        args[0],
+		Args:       args[1:],
+		Env:        cfg.EnvList(),
+		User:       cfg.User,
+		TaskDir:    cfg.TaskDir().Dir,
+		StdoutPath: cfg.StdoutPath,
+		StderrPath: cfg.StderrPath,
+	}
+	ps, err := execImpl.Launch(execCmd)
+	if err != nil {
+		pluginClient.Kill()
+		return nil, nil, err
+	}
+	d.logger.Debug("started new QemuVM", "ID", vmID)
+
+	//TODO(preetha) figure out if monitor path is needed
```

Comment:
Yes. It's used in StopTask: https://github.com/hashicorp/nomad/pull/4800/files#diff-623238b3f3bd24423a54a318f28a0fbfR482